Updated Lift Spec (1/x)

matthewmueller opened the pull request:
[Rendered View](https://github.com/prisma/specs/blob/lift/lift/Readme.md)

Rewrote the Lift spec to hopefully unblock planning. Let me know if you want me to flush out some more specific areas. 

There's still lots of TODOs, I'm not as intimately familiar with Lift as @timsuchanek and @mavilein, but hopefully this will give us a basis to discuss improvements going forward.

Part of https://github.com/prisma/specs/issues/156

janpio commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
-As mentioned in [Declarative migration](#how-does-naming-of-migrations-work), the user adjusts the `datamodel.prisma` file, which results in a new migration
-folder when running `prisma migrate`. In order for the migration engine to know which changes need to be applied, it needs a "copy" of the datamodel. As this
-"copy" describes the datamodel in a specific point in time, we also refer to it as a "snapshot".
+### Hook
```
Do we actually, really already have Migration Hooks?

mavilein replied:
No

janpio replied:
This is now covered: https://github.com/prisma/specs/blob/lift/lift/Readme.md#hook

janpio commented on lift/Readme.md:
````diff
@@ -6,271 +6,360 @@
-- Prisma can ensure, that migrations don't get run twice
-- Prisma can give an overview in a GUI interface about the currently running migration and migrations that already have been running
-- Prisma can rollback migrations when there has been an error
+```
````
Cosmetics: This diagram is a bit messed up on GH. Is this fixable?

matthewmueller replied:
<img width="552" alt="Screen Shot 2019-09-27 at 11 06 38 AM" src="https://user-images.githubusercontent.com/170299/65757166-e50ee700-e116-11e9-89fb-e9a9e8dd7a84.png">

^ seems okay to me.

janpio commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
-When a migration is running or has been executed successfully, Prisma writes this information into a table or collection in the database. This table is also
-known as the "Migration History". It includes all successful and failed migrations of the specific Prisma Project. The benefits of storing this migration
-history are the following:
+Lift has 2 parts: a **Lift Client** and a **Lift Server**. By default, the Lift Server runs locally as a sidecar process. This makes it easy to get started. As
```
Valid to link "sidecar process" to e.g. https://blog.davemdavis.net/2018/03/13/the-sidecar-pattern/ for context / noobs like me?

janpio commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
-known as the "Migration History". It includes all successful and failed migrations of the specific Prisma Project. The benefits of storing this migration
-history are the following:
+Lift has 2 parts: a **Lift Client** and a **Lift Server**. By default, the Lift Server runs locally as a sidecar process. This makes it easy to get started. As
+your team's data requirements become more complex, you may prefer to handle your migrations on a remote machine where you have fine-grained access control over
```
Do we already support remote migration servers?

mavilein replied:
No

janpio replied:
Currently still present in first paragraph here: https://github.com/prisma/specs/blob/lift/lift/Readme.md#architecture If we want to turn the "remote" stuff into a future thing, then it maybe should be removed here as well.

janpio replied:
Now commented at https://github.com/prisma/specs/pull/159/files#r328962349

janpio commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
 
-### Naming of migration folders
+- **steps.json:** contains a JSON list of steps to run against the database
+- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time.
```
before or after the migration?

mavilein replied:
It is the schema you used when calling `prisma lift save`. So it is the state after the migration.

janpio replied:
Fixed: https://github.com/prisma/specs/blob/lift/lift/Readme.md#save

janpio commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
+    └─ steps.json
+    └─ schema.prisma
+    └─ README.md
+    └─ before.sh
```
`before.sh` and `after.sh` are a new concept here (or did I miss it?). I assume those are hooks? If so, and these are the common names for them, this should be explained in the "Hooks" sections.

matthewmueller replied:
good call, it's defined now

janpio commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
+A migration contains 3 files:
 
-### Naming of migration folders
+- **steps.json:** contains a JSON list of steps to run against the database
```
Both the up and down versions? Or are those the sames, and the different result is just calculated when executing up or down?

mavilein replied:
They are the same. We calculate up and down on the fly.

janpio commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
-- Making sure that the data adheres to constraints that you want to introduce, like `unique` or `non null` constraints
-- Adding specific database primitives like functions in Postgres, which are not yet supported by Prisma
-- Initializing data after a migration has been executed. Both SQL or the Prisma Client could be useful here
+The Lift server is a low-level interface that the Lift Client communicates with. Currently the client communicates to the migration server in the JSONRPC format
```
migration server = lift server?

mavilein replied:
Should we actually call those servers? Or should we rather embrace our internal engine term?

janpio commented on lift/Readme.md:
````diff
@@ -6,271 +6,360 @@
 
-```bash
-prisma migrate                                Creates a migration, clears the draft and migrates the database
+Save method takes the difference of your schema file with the current state of your datasources and creates a migration folder. This migration folder contains
````
migration vs. migration folder. Is it the same thing? Maybe define.

mavilein commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
+### Migration
 
-### Datamodel Snapshots
+A Migration is a grouping of one or more Steps. Migrations run in a transaction if the datasource allows it.
```
This should be marked as not implemented.

mavilein commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
+### Hook
 
-### Migration Engine
+A hook is a custom shell script that runs either before or after the migration. Hooks give you full control over your migrations. You can write hooks to:
```
It does not really give you full control imo. I would phrase it that it allows you to customise a migration when the normal migrations don't cut it.

mavilein commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
+#### unapplyMigration
 
-### Up & Down in migration scripts
+Unapplies the Steps in the migrations folder. This is called by the Down method in the Lift Client.
```
This is not entirely correct. `unapplyMigration` just rolls back the last migration in the database. This method is called repeatedly by the CLI.

matthewmueller replied:
good call

mavilein commented on lift/Readme.md:
```diff
@@ -6,271 +6,360 @@
+### Step
 
-Assuming that the reader knows, what Prisma 1.x looks like, the following terms are new concepts which we introduce with this spec.
+A Step is a single command that we'll run against the database. Steps describe models, fields and relationships generically, so they can be easily translated to
```
This sounds like as if one step would always lead to one command against the database. But this is not the case. Sometimes a step does not even trigger a single database command, e.g. `CreateEnum` which does not have a representation in the database at the moment. Often multiple steps are grouped into one command, e.g. a `CreateModel` and multiple `CreateField` result in one `CREATE TABLE` in the sql connector.

janpio commented:
A list of things that were present in the past, but are not any more:

1. Imperative vs. Declarative Migration: Pretty much fluff, not really necessary.
2. Migration Engine: Technical implementation detail?
3. Migration History: Seems like this could be important.
4. Migration scripts: Now called hooks? Missing a few details.

janpio commented:
Thought: Current `prisma2 dev` might also be considered part of Lift and might be relevant here.

matthewmueller commented:
Thanks for reviewing! I've incorporated the feedback. 

@sorenbs, good point I've also replaced that section with 

> ### How can you rename a model in Lift?
> 
> To rename a field, you'll add an `UpdateModel` step in your **steps.json** file directly.

janpio commented on lift/Readme.md:
```diff
@@ -245,9 +253,11 @@ migrations/
 
-- **steps.json:** contains a JSON list of steps to run against the database
-- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time.
+- **steps.json:** contains a JSON list of steps to run against the database. Steps contains only the up steps, the down steps are calculated on the fly.
```
Is that technically correct @mavilein? Or is this more general and both steps are generated?

mavilein replied:
It's a little bit more involved. The `steps.json` only represents the changes that should be applied to the current datamodel to deduct the desired datamodel. Based on the current and desired datamodel we generate the actual database steps that will be performed for up and down.
To put it into equations:
1. `desired_datamodel = current_datamodel + steps.json`
2. `actual_database_steps = generate(current_datamodel, desired_datamodel)` <- this function is connector specific

Does that explanation help?

janpio replied:
It does and gives me details, thanks! And it also kind of confirms that "up steps" is valid enough here in what @matthewmueller wrote - it contains things to go from current to desired, which I think is fine to describe as "up steps", right?

mavilein replied:
Yes i think it's accurate enough.

janpio commented on lift/Readme.md:
```diff
@@ -245,9 +253,11 @@ migrations/
+- **steps.json:** contains a JSON list of steps to run against the database. Steps contains only the up steps, the down steps are calculated on the fly.
+- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time after the migration has occurred.
 - **README.md:** contains information about the migration. Includes the underlying raw commands (e.g. SQL) that run against the datasource.
+- **before.sh:** optional hook you can run before migrating your schema.
```
Will this be limited to `.sh` files or will it also enable `.js` and similar?

janpio commented on lift/Readme.md:
```diff
@@ -2,275 +2,378 @@
+- **steps.json:** contains a JSON list of steps to run against the database. Steps contains only the up steps, the down steps are calculated on the fly.
+- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time after the migration has occurred.
+- **README.md:** contains information about the migration. Includes the underlying raw commands (e.g. SQL) that run against the datasource.
+- **before.sh:** optional hook you can run before migrating your schema.
```
```suggestion
- **before.sh:** optional [hook](#hook) you can run before migrating your schema.
```

janpio commented on lift/Readme.md:
```diff
@@ -2,275 +2,378 @@
+- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time after the migration has occurred.
+- **README.md:** contains information about the migration. Includes the underlying raw commands (e.g. SQL) that run against the datasource.
+- **before.sh:** optional hook you can run before migrating your schema.
+- **after.sh:** optional hook you can run after migrating your schema.
```
```suggestion
- **after.sh:** optional [hook](#hook) you can run after migrating your schema.
```

janpio commented on lift/Readme.md:
```diff
@@ -2,275 +2,378 @@
 
-When executing `prisma migrate` without providing any args, a migration with the name `yyyyMMddHHmmss` (e.g. `20190322092247`) will be created. Especially when
-working in a team, it can be very useful to add a custom name to the migration. A name can be provided like this:
+**Note:** `migrate save` does not run migrations, it simply creates them.
```
What does `migrate save` refer to here? CLI command?

janpio commented on lift/Readme.md:
```diff
@@ -2,275 +2,378 @@
-As mentioned in the [basic example](#basic-example), a simple `prisma migrate` is needed to migrate the database and create a new migration.
+Lift has 2 parts: a **Lift Client** and a **Lift Engine**. By default, the Lift Engine runs locally as a
+[sidecar process](https://blog.davemdavis.net/2018/03/13/the-sidecar-pattern/). This makes it easy to get started. As your team's data requirements become more
+complex, you may prefer to handle your migrations on a remote machine where you have fine-grained access control over migrations.
```
"remote machine" / remote migration engine is not implemented yet and was removed elsewhere in the spec.

janpio commented:
As we talked about the actual RPC commands under "Lift Engine" should be moved to a new issue that talks about making these commands a public API that should be stable. It would probably also make sense to have these in a `.md` file in `prisma/lift` - I am sure @tomhoule can help out there.

janpio commented on lift/Readme.md:
```diff
@@ -2,275 +2,424 @@
+
+#### `prisma2 dev`
+
+TODO
```
Can someone suggest a wording of what the Lift parts of `prisma2 dev` actually are and how they work?

janpio replied:
Ping @timsuchanek @tomhoule

janpio commented:
What was the motivation of replacing the generalized spec of what the CLI should do to pasting the documentation? Isn't that the wrong way around?

PS: Content of the last two commits is kind of messed up :/
